File: dynamicprogramming/knapsack/knapsack.py
# This algorithm doesn't tell you what was placed in the bag, only its maximum
# value.  To get the items in the list, you can store whether the maximum for
# the call included the current item.

from json import dumps
from random import randint

MINWEIGHT = 1
MAXWEIGHT = 100
MINVALUE = 10
MAXVALUE = 1000

values = []
weights = []

def knapsack(item, remaining):
    """Returns the maximum value; but not what is in the knapsack."""

    global values
    global weights

    # last case where it's called, no value, no weight cost.
    if item == len(values):
        return 0

    if weights[item] > remaining:
        return knapsack(item+1, remaining)
    else:
        return max(knapsack(item+1, remaining),
                   knapsack(item+1, remaining-weights[item]) + values[item])

def main():
    global values
    global weights
    global included

    for i in range(10):
        weights.append(randint(MINWEIGHT, MAXWEIGHT))
    for i in range(10):
        values.append(randint(MINVALUE, MAXVALUE))

    knapsack_weight = 250
    maximum_value = knapsack(0, knapsack_weight)
    print "maximum value: %d" % maximum_value
    print "weights: %s" % dumps(weights)
    print "values: %s" % dumps(values)
    
if __name__ == "__main__":
    main()

File: dynamicprogramming/knapsack/knapsack.c

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <time.h>

#define NUM_ELEMENTS(X) (sizeof(X)/sizeof(*X))
#define MAX(X, Y) ((X) > (Y)) ? (X) : (Y)

#define NUMITEMS 10
#define KNAPSACK_WEIGHT 250

#define MINWEIGHT 1
#define MAXWEIGHT 100
#define MINVALUE 10
#define MAXVALUE 1000

static int *values = NULL;
static int *weights = NULL;

static int knapsack(int item, int remaining) {

    if (item == NUMITEMS) {
        return 0;
    }

    if (weights[item] > remaining) {
        return knapsack(item+1, remaining);
    } else {
        return MAX(knapsack(item+1, remaining),
                   knapsack(item+1, remaining-weights[item]) + values[item]);
    }
}

static int ranged_random(int min, int max) {
    int x = random();
    x %= max;
    if (x < min) {
        x = min; /* not ideal. but whatever. :D */
    }

    return x;
}

int main(void)
{
    int i, max;
    values = malloc(sizeof(int) * NUMITEMS);
    weights = malloc(sizeof(int) * NUMITEMS);

    /* srand is the normal unix, but on darwin, it's srandom */
    srandom(time(NULL));

    for (i = 0; i < NUMITEMS; i++) {
        values[i] = ranged_random(MINVALUE, MAXVALUE);
        weights[i] = ranged_random(MINWEIGHT, MAXWEIGHT);
    }

    max = knapsack(0, KNAPSACK_WEIGHT);

    printf("maximum knapsack value: %d\n", max);

    free(values);
    free(weights);

    return 0;
}
